src/dag/dag.ml

```diff
@@ -0,0 +1,152 @@
+open! Stdune
+
+include Dag_intf
+
+module Make(Value : Value) : S with type value := Value.t = struct
+  type t = {
+    mutable num : int;
+    mutable index : int;
+    mutable arcs : int;
+  }
+
+  type node_info = {
+    id : int;
+    mutable index : int;
+    mutable level : int;
+    mutable deps : node list;
+    mutable rev_deps : node list;
+  }
+
+  and node =
+    { data : Value.t
+    ; info : node_info
+    }
+
+  exception Cycle of node list
+
+  let create () = {
+    num = 0;
+    index = 0;
+    arcs = 0;
+  }
+
+  let gen_index (dag : t) =
+    dag.index <- dag.index - 1;
+    dag.index
+
+  let create_node_info dag =
+    dag.num <- dag.num + 1;
+    {
+      id = dag.num;
+      index = gen_index dag;
+      level = 1;
+      deps = [];
+      rev_deps = [];
+    }
+
+  let delta dag =
+    let n = dag.num in
+    let m = dag.arcs in
+    let delta = min (float m ** (1.0 /. 2.0)) (float n ** (2.0 /. 3.0)) |> int_of_float in
+    delta
+
+  let add dag v w =
+    let delta = delta dag in
+
+    dag.arcs <- dag.arcs + 1;
+
+    (* Printf.printf "new add %s -> %s\n%!" v.input w.input; *)
+    let marked_ids = ref Int.Set.empty in
+    let arcs = ref 0 in
+    let f = ref [] in
+    let b = ref [] in
+
+    let rec bvisit y acc =
+      marked_ids := Int.Set.union !marked_ids (Int.Set.singleton y.info.id);
+      if List.exists y.info.rev_deps ~f:(fun x -> btraverse x y (x :: acc)) |> not then begin
+        b := List.append !b [y];
+        false end
+      else
+        true
+    and btraverse x _y acc =
+      if x.info.id = w.info.id then raise (Cycle acc);
+      arcs := !arcs + 1;
+      if !arcs >= delta then begin
+        w.info.level <- v.info.level + 1;
+        w.info.rev_deps <- [];
+        b := [];
+        true
+      end else begin
+        if Int.Set.mem !marked_ids x.info.id |> not then
```

Comment:
the `|> not` is a bit hard to see, I'd suggest to invert the branches. (this can be rewritten as `mem ... && bvisit ...` but since it is effectful, that does not seem to be a good idea)

Reply:
Agreed, I inverted the branches